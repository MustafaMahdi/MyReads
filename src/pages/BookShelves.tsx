import { Book as BookInterface } from "../models/book";
import Book from "../components/Book";
import { ShelvesType } from "../defines/defines";
import { Link } from 'react-router-dom';


const BookShelves: React.FC<{
  books: BookInterface[],
  changeShelfHandler: (book: BookInterface, selectedShelf: string) => void
}> = (props) => {

  const currentlyReading = props.books.filter(book => book.shelf === ShelvesType.CurrentlyReading)
    .map(book => (
      <li key={book.id} >
        <Book
          book={book}
          currentShelf={ShelvesType.CurrentlyReading}
          onChangeShelf={props.changeShelfHandler}
        />
      </li>
    ));

  const wantToRead = props.books.filter(book => book.shelf === ShelvesType.WantToRead)
    .map(book => (
      <li key={book.id} >
        <Book
          book={book}
          currentShelf={ShelvesType.WantToRead}
          onChangeShelf={props.changeShelfHandler}
        />
      </li>
    ));

  const read = props.books.filter(book => book.shelf === ShelvesType.Read)
    .map(book => (
      <li key={book.id} >
        <Book
          book={book}
          currentShelf={ShelvesType.Read}
          onChangeShelf={props.changeShelfHandler}
        />
      </li>
    ));
  return (
    <div>
      <div className="list-books-title">
        <h1>MyReads: A Book Tracking App</h1>
      </div>
      <div className="list-books-content">
        <div>
          <div className="bookshelf">
            <h2 data-testid="currentlyReading" className="bookshelf-title">Currently Reading</h2>
            <div className="bookshelf-books">
              <ol className="books-grid">
                {
                  currentlyReading
                }
              </ol>
            </div>
          </div>
          <div className="bookshelf">
            <h2 data-testid="wantToRead" className="bookshelf-title">Want to Read</h2>
            <div className="bookshelf-books">
              <ol className="books-grid">
                {
                  wantToRead
                }
              </ol>
            </div>
          </div>
          <div className="bookshelf">
            <h2 data-testid="read" className="bookshelf-title">Read</h2>
            <div className="bookshelf-books">
              <ol className="books-grid">
                {
                  read
                }
              </ol>
            </div>
          </div>
        </div>
      </div>
      <div className="open-search">
        <Link to="/search">Add a book</Link>
      </div>
    </div>
  );
};

export default BookShelves;


