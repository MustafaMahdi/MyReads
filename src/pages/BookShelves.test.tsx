import React from 'react';
import { render, screen } from '@testing-library/react';
import BookShelves from './BookShelves';
import { ShelvesType } from '../defines/defines';
import { Book as BookInterface } from "../models/book";
import { BrowserRouter, Link } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Book from '../components/Book';

describe('BookShelves component', () => {
  const props = {
    books: [{
      authors: ['Thomas HelwysRichard'],
      title: 'A Short Declaration of the Mystery of Iniquity',
      shelf: ShelvesType.None, id: "id",
      imageLinks: {
        thumbnail: "http://books.google.com/books/content?id=sJf1vQAACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api"
      }
    }],
    changeShelfHandler: (book: BookInterface, selectedShelf: string) => { },
  }
  it('renders BookShelves Component', () => {
    render(<BrowserRouter><BookShelves {...props} /></BrowserRouter>);
    expect(screen.getByText(/MyReads: A Book Tracking App/i)).toBeVisible();


  });
  it('should render Currently Reading', () => {
    render(<BrowserRouter><BookShelves {...props} /></BrowserRouter>);
    expect(screen.getByTestId('currentlyReading')).toBeVisible();
  });
  it('should render Want to read', () => {
    render(<BrowserRouter><BookShelves {...props} /></BrowserRouter>);
    expect(screen.getByTestId('wantToRead')).toBeVisible();
  });
  it('should render Read', () => {
    render(<BrowserRouter><BookShelves {...props} /></BrowserRouter>);
    expect(screen.getByTestId('read')).toBeVisible();
  });
  it('should route to search page when click on the link action', () => {
    const component = renderer.create(
      <BrowserRouter><Link to="/search" /></BrowserRouter>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
  it('should render Book Component inside BookShelves component by check the title of book which rendered at the Book component', () => {
    render(<BrowserRouter><BookShelves {...props} /></BrowserRouter>);
    render(<Book book={props.books[0]} currentShelf={ShelvesType.CurrentlyReading} onChangeShelf={props.changeShelfHandler} />);
    expect(screen.getByText(/A Short Declaration of the Mystery of Iniquity/i)).toBeVisible();
  })
});

