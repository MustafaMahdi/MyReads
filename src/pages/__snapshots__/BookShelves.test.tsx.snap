// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookShelves component should route to search page when click on the link action 1`] = `
<a
  href="/search"
  onClick={[Function]}
/>
`;
